airflow/sensors/smart_sensor_operator.py

```diff
@@ -0,0 +1,730 @@
+    """
+    Smart sensor operators are derived from this class.
+
+    Smart Sensor operators keep refresh a dictionary by visiting DB.
+    Taking qualified active sensor tasks. Different from sensor operator,
+    Smart sensor operators poke for all sensor tasks in the dictionary at
+    a time interval. When a criteria is met or fail by time out, it update
+    all sensor task state in task_instance table
+
+    :param soft_fail: Set to true to mark the task as SKIPPED on failure
+    :type soft_fail: bool
+    :param poke_interval: Time in seconds that the job should wait in
+        between each tries
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+    :type mode: str
+    """
+    ui_color = '#e6f1f2'
+
+    @apply_defaults
+    def __init__(self,
+                 poke_interval=180,
+                 smart_sensor_timeout=60 * 60 * 24 * 7,
+                 soft_fail=False,
+                 shard_min=0,
+                 shard_max=100000,
+                 poke_exception_cache_ttl=600,
+                 poke_timeout=6,
+                 poke_exception_to_fail_task_threshold=3,
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        # super(SmartSensorOperator, self).__init__(*args, **kwargs)
+        self.poke_interval = poke_interval
+        self.soft_fail = soft_fail
+        self.timeout = smart_sensor_timeout
+        self._validate_input_values()
+        self.hostname = ""
+
+        self.sensor_works = []
+        self.cached_dedup_works = {}
+        self.cached_sensor_exceptions = {}
+
+        self.max_tis_per_query = 50
+        self.shard_min = shard_min
+        self.shard_max = shard_max
+        self.poke_exception_cache_ttl = poke_exception_cache_ttl
+        self.poke_timeout = poke_timeout
+        self._poke_exception_to_fail_task_threshold = poke_exception_to_fail_task_threshold
+
+    def _validate_input_values(self):
+        if not isinstance(self.poke_interval, (int, float)) or self.poke_interval < 0:
+            raise AirflowException(
+                "The poke_interval must be a non-negative number")
+        if not isinstance(self.timeout, (int, float)) or self.timeout < 0:
+            raise AirflowException(
+                "The timeout must be a non-negative number")
+
+    @provide_session
+    def _load_sensor_works(self, session=None):
+        """
+        Refresh sensor instances need to be handled by this operator. Put the context,
+        hashcode and sensor instance start_date in a wrapper class
+        :param session:
+        :return:
+        """
+        SI = SensorInstance
+        start_query_time = time.time()
+        query = session.query(SI) \
+            .filter(SI.state == State.SENSING)\
+            .filter(SI.shardcode < self.shard_max,
+                    SI.shardcode >= self.shard_min)
+        tis = query.all()
+        session.commit()
```

Comment:
Do we need this? We are simply querying in this function so I think we can safely remove this